db/tables.py

```diff
@@ -408,4 +413,46 @@ def infer_table_column_types(
             table_name,
             column_name,
             engine,
+            conn=conn,
         )
+
+
+class CreateTempTableAs(DDLElement):
+    def __init__(self, name, selectable):
+        self.name = name
+        self.selectable = selectable
+
+
+@compiler.compiles(CreateTempTableAs)
+def compile(element, compiler, **_):
+    return "CREATE TEMPORARY TABLE %s AS (%s)" % (
+        element.name,
+        compiler.sql_compiler.process(element.selectable, literal_binds=True),
+    )
+
+
+def infer_table_column_types(schema, table_name, engine):
+    table = reflect_table(table_name, schema, engine)
+    temp_name = "temp_table"
+
+    select_table = select(table)
+    with engine.connect() as conn:
+        with conn.begin():
+            conn.execute(CreateTempTableAs(temp_name, select_table))
+        with conn.begin():
+            temp_table = reflect_table(temp_name, None, conn)
+        try:
+            update_table_column_types(
+                None, temp_table.name, engine, table=temp_table, conn=conn
+            )
+            with conn.begin():
+                temp_table = reflect_table(temp_name, None, conn)
+                types = [c.type.__class__ for c in temp_table.columns]
+        except Exception as e:
+            # Ensure the temp table is deleted
+            with conn.begin():
+                temp_table.drop()
+            raise e
+        with conn.begin():
+            temp_table.drop()
```

Comment:
Same as above.  If we can't get the temp table to work (since SQLAlchemy won't give up connections), we should probably just use a normal table to keep from confusing people.